### ISSUE TITLE
Passing OData parameters to GET collection methods in API and ODATA Controllers 

### ISSUE THREAD

**Issue by stratdaz** [RicoSuter/NSwag#2202]
Using the Typescript Client in NSwagStudio, my "queryenabled" API and ODATA controller parameterless GET  collection methods  are generated having no parameter to facilitate passing OData parameters such as $filter, $expand, etc.

Is there an option to have these methods accept a parameter or is it a case of modifying the Typescript post generation?

**Comment by dis081278** [comment]
Hi, I am interested in a reply to this question as well.  I'm generating c# clients in NSwagStudio and have the same question.

**Comment by jeremyVignelles** [comment]
I don't understand the question here. Do you have a concrete example of what you want and what you expect NSwag to do ?

**Comment by stratdaz** [comment]
Hi,

I may be misunderstanding but this is what I was referring to ...

My Web API controllers GETs are query enabled. I used paged grids, for example, in my web application so a call to the server GET method from the client for a page might look something like ...

**http://host/MyWeb.Api/api/WorkboardPaths?$format=json&$top=10&$skip=20&%count=true**

My Typescript client generated from my Web API for the WorkboardPaths controller looks like this ...

   **_constructor(@Inject(HttpClient) http: HttpClient, @Optional() @Inject(API_BASE_URL) baseUrl?: string) {
        this.http = http;
        this.baseUrl = baseUrl ? baseUrl : "";
    }

    getWorkboardPaths(): Observable<WorkboardPath[] | null> {
        let url_ = this.baseUrl + "/api/WorkboardPaths";
        url_ = url_.replace(/[?&]$/, "");

        let options_ : any = {
            observe: "response",
            responseType: "blob",
            headers: new HttpHeaders({
                "Accept": "application/json"
            })
        };

        return this.http.request("get", url_, options_).pipe(_observableMergeMap((response_ : any) => {
            return this.processGetWorkboardPaths(response_);
        })).pipe(_observableCatch((response_: any) => {
            if (response_ instanceof HttpResponseBase) {
                try {
                    return this.processGetWorkboardPaths(<any>response_);
                } catch (e) {
                    return <Observable<WorkboardPath[] | null>><any>_observableThrow(e);
                }
            } else
                return <Observable<WorkboardPath[] | null>><any>_observableThrow(response_);
        }));
    }_**

I guess I'm wondering how to pass the odata parameters like $top, $skip to the getWorkboardPaths() method above?

**Comment by dis081278** [comment]
Yes, that's what I'm wondering as well.  My Get Action is as follows:

        [EnableQuery]
        [HttpGet, Route("")]
        [ResponseType(typeof(DbSet<Administrator>))]
        public IQueryable<Book> GetBooks()
        {
            return unitOfWork.Books.AsQueryable();
        }


The client generated has the following signature: 
        
System.Threading.Tasks.Task<DbSetOfBooks> GetBooksAsync();
 
How do I pass the OData query to this method?

**Comment by jeremyVignelles** [comment]
I think that this is a question for ASP.net core's ApiExplorer.

NSwag only get what's returned by the ApiExplorer. You could also use a IDocumentProcessor to modify  the generates document and add the OData query paramters

**Comment by stratdaz** [comment]
Perhaps something to do with the fact that the get methods are parameter-less. 

And explicitly adding an ODataQueryOptions parameter might work?

public IHttpActionResult Get(**ODataQueryOptions**<TEntity> **queryOptions**)
{
  QueryOptionsUtilities.RemoveExpandDetails(queryOptions);
 
  // ... 
}

**Comment by figuerres** [comment]
this problem has been getting passed around like a hot potato for ever..... Microsoft uses Odata but does not help with the swagger support for it.  swagger generators say they do not want to handle it they want the OData folks to fix it.... the OData folks say it works find just make the right swagger generator and around it goes again....
